src/framework/ui/internal/uiarrangement.cpp

```diff
@@ -0,0 +1,168 @@
+        updateData(DataKey::UiStates, m_states, m_statesNotifications);
+        updateData(DataKey::UiToolConfigs, m_toolconfigs, m_toolconfigsNotifications);
+    });
+
+    updateData(DataKey::UiSettings, m_settings, m_valuesNotifications);
+    updateData(DataKey::UiStates, m_states, m_statesNotifications);
+    updateData(DataKey::UiToolConfigs, m_toolconfigs, m_toolconfigsNotifications);
+}
+
+void UiArrangement::updateData(DataKey key, QJsonObject& obj, Notifications& notifications) const
+{
+    RetVal<QByteArray> data = workspacesDataProvider()->rawData(key);
+    if (!data.ret) {
+        LOGD() << "no data: " << key_to_string(key);
+        obj = QJsonObject();
+        return;
+    }
+
+    QJsonObject oldObj = obj;
+    obj = QJsonDocument::fromJson(data.val).object();
+
+    Notifications::Iterator it = notifications.begin();
+    for (; it != notifications.end(); ++it) {
+        const QString& key = it.key();
+
+        if (oldObj.value(key) != obj.value(key)) {
+            it.value().notify();
+        }
+    }
+}
+
+void UiArrangement::saveData(workspace::DataKey key, const QJsonObject& obj)
+{
+    QByteArray data = QJsonDocument(obj).toJson();
+
+    workspacesDataProvider()->setRawData(key, data);
+}
+
+QString UiArrangement::value(const QString& key) const
+{
+    QJsonValue val = m_settings.value(key);
+    return val.toString();
+}
+
+void UiArrangement::setValue(const QString& key, const QString& val)
+{
+    m_settings[key] = val;
+    saveData(DataKey::UiSettings, m_settings);
+    if (m_valuesNotifications.contains(key)) {
+        m_valuesNotifications[key].notify();
+    }
+}
+
+mu::async::Notification UiArrangement::valueChanged(const QString& key) const
+{
+    return m_valuesNotifications[key];
+}
+
+QByteArray UiArrangement::state(const QString& key) const
+{
+    QJsonValue val = m_states.value(key);
+    QString valStr = val.toString();
+    return valStr.toLocal8Bit();
+}
+
+void UiArrangement::setState(const QString& key, const QByteArray& data)
+{
+    m_states[key] = QString::fromLocal8Bit(data);
+    saveData(DataKey::UiStates, m_states);
+    if (m_statesNotifications.contains(key)) {
+        m_statesNotifications[key].notify();
+    }
+}
+
+mu::async::Notification UiArrangement::stateChanged(const QString& key) const
+{
+    return m_statesNotifications[key];
+}
+
+ToolConfig UiArrangement::toolConfig(const QString& toolName) const
+{
+    TRACEFUNC;
+
+    //! NOTE Maybe we need to cache?
+
+    QJsonObject confObj = m_toolconfigs.value(toolName).toObject();
+    QJsonArray itemsArr = confObj.value("items").toArray();
+
+    ToolConfig config;
+
+    config.items.reserve(itemsArr.size());
+    for (const QJsonValue& v : itemsArr) {
+        QJsonObject itemObj = v.toObject();
+
+        ToolConfig::Item item;
+        item.action = itemObj.value("action").toString().toStdString();
+        item.show = itemObj.value("show").toInt(1);
```

Comment:
toInt(1) - Is it the setting of the default value? Then it is wrong

Reply:
Yes, this is the default.
Why are you thinking wrong?
This should not be the case, the value should always be set, but if there was some kind of software error, then I prefer that all the controls are shown, instead of all are hidden

Reply:
Sorry, my mistake. I looked at the description of the QVariant::toInt method instead of QJsonObject::toInt